### ISSUE TITLE
trie in variable-length lookbehind is not anchored

### ISSUE THREAD

**Issue by hvds** [Perl/perl5#19168]
```
% perl -v | head -2 | tail -1
This is perl 5, version 34, subversion 0 (v5.34.0) built for x86_64-linux

# this correctly fails to match
% perl -E 'no warnings qw{experimental}; "abfoo" =~ /(?=foo)(?<=(a|a))/ and say $1'

# this wrongly matches: it should only match if we ask for `ab`
% perl -E 'no warnings qw{experimental}; "abfoo" =~ /(?=foo)(?<=(a|ac))/ and say $1'
a
```

This should not match; my guess is that the trie code is unaware of whatever we've done to anchor the end of the lookbehind.

I won't get a chance to look at this for a few days: @khwilliamson @demerphq any ideas?

The original work was in 2fe8bdb, I haven't checked if there were later updates.

**Comment by hvds** [comment]
I spent a little time looking at this, but the issue appears to be in the part of regexec.c that I've not understood since it was made iterative in 2006 (the 3600-odd line function S_regmatch). I believe the pattern is compiled correctly.

Note that the `/(?=foo)(?<=(a|ac))/` example above actually extracts the prefix to make `/(?=foo)(?<=a(|c))/`, but the zero-length match in the trie is not the issue: the same occurs with the unprefixed `/(?=foo)(?<=(a|bc))/`.

**Comment by hvds** [comment]
@demerphq any chance you have headspace to look at this one? It's possible that the trie code and the vlb code just need to be a little more aware of each other, but it may go deeper than that.

**Comment by demerphq** [comment]
On Fri, 18 Feb 2022, 23:16 Hugo van der Sanden, ***@***.***>
wrote:

> @demerphq <https://github.com/demerphq> any chance you have headspace to
> look at this one? It's possible that the trie code and the vlb code just
> need to be a little more aware of each other, but it may go deeper than that
>

Sure thing, I'll take a look.

Yves

>

**Comment by demerphq** [comment]
On Sat, 19 Feb 2022 at 02:29, demerphq ***@***.***> wrote:

>
>
> On Fri, 18 Feb 2022, 23:16 Hugo van der Sanden, ***@***.***>
> wrote:
>
>> @demerphq <https://github.com/demerphq> any chance you have headspace to
>> look at this one? It's possible that the trie code and the vlb code just
>> need to be a little more aware of each other, but it may go deeper than that
>>
>
> Sure thing, I'll take a look.
>

 The only thing that we can do is disable trie optimization inside of
lookbehind.

(?<=a|bc)

is not really an "or", it is an *and*, the pattern is equivalent to

(?<=a)(?<=bc)

the key point being that (?<=a) asserts that the "immediately preceding
character cannot be 'a' and the (?<=bc) asserts that the "immediately two
preceding character cannot be 'bc'.  More specifically it is checking two
different offsets.

In short this cannot be processed by a trie. I actually find it
questionable if it should be allowed at all. I think it wasn't once upon a
time. (When the trie optimization was implemented.) Anyway, for now the
only solution is to teach regcomp.c to not trie inside of lookbehind, or to
perhaps not trie if the keys are different lengths, probably the former is
easier to implement than the latter so that is what I will do.

Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by hvds** [comment]
> the key point being that (?<=a) asserts that the "immediately preceding character cannot be 'a'

Am I misreading this? I'd expect that to be the meaning of `(?<!a)`, negative lookbehind.

> I actually find it questionable if it should be allowed at all. I think it wasn't once upon a time.

Experimental support for variable-length lookbehind was introduced in 5.30, it was a fatal error before that. I stumbled on this issue when Ricardo proposed making it non-experimental.

> Anyway, for now the only solution is to teach regcomp.c to not trie inside of lookbehind, or to perhaps not trie if the keys are different lengths, probably the former is easier to implement than the latter so that is what I will do.

I don't think my own use of lookbehinds would lose much from the lack of trie-optimization - I rarely have anything particularly complex in there - but I don't know how other people's usage would stack up. But it would seem a shame if a side-effect of this experimental feature is that we have to pessimize what we supported before - @rjbs, @khwilliamson do you have thoughts on this?

**Comment by demerphq** [comment]
On Sun, 20 Feb 2022 at 02:16, demerphq ***@***.***> wrote:

> On Sat, 19 Feb 2022 at 02:29, demerphq ***@***.***> wrote:
>
>>
>>
>> On Fri, 18 Feb 2022, 23:16 Hugo van der Sanden, ***@***.***>
>> wrote:
>>
>>> @demerphq <https://github.com/demerphq> any chance you have headspace
>>> to look at this one? It's possible that the trie code and the vlb code just
>>> need to be a little more aware of each other, but it may go deeper than that
>>>
>>
>> Sure thing, I'll take a look.
>>
>
>  The only thing that we can do is disable trie optimization inside of
> lookbehind.
>
> (?<=a|bc)
>
> is not really an "or", it is an *and*, the pattern is equivalent to
>
> (?<=a)(?<=bc)
>
> the key point being that (?<=a) asserts that the "immediately preceding
> character cannot be 'a' and the (?<=bc) asserts that the "immediately two
> preceding character cannot be 'bc'.  More specifically it is checking two
> different offsets.
>
> In short this cannot be processed by a trie. I actually find it
> questionable if it should be allowed at all. I think it wasn't once upon a
> time. (When the trie optimization was implemented.) Anyway, for now the
> only solution is to teach regcomp.c to not trie inside of lookbehind, or to
> perhaps not trie if the keys are different lengths, probably the former is
> easier to implement than the latter so that is what I will do.
>

Turns out this is just broken, with or without the trie optimization.

./perl -Ilib -Mre=debug -E 'no warnings qw{experimental}; "abfoo" =~
/(?=foo)(?<=(a|ac))/ and say $1'

the problem is that it is trying to match "a" at offset -2, which is only
suitable for matching "ac". And it does match, so the alternation succeeds
and the pattern inappropriately matches. Same problem with the trie logic.
With the untried form it is at least somewhat possible to say "this exact
must match at the offset of its length", but currently that does not happen.

Karl added support for this logic with I believe the intention of
supporting unicode case folding scenarios where the folding causes what
looks like a fixed length pattern to match different folded lengths for
instance (?<=\xDF) under unicode should match \xDF and 'ss'. I believe he
has handled the edge cases this creates. However I do not believe that
there is any support for making alternation work.

I am looking into what I can do, maybe I can find a way to make this work
properly. We will see. I am a little skeptical about this feature and if we
should support it.

cheers,
Yves




-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
On Sun, 20 Feb 2022 at 05:41, Hugo van der Sanden ***@***.***>
wrote:

> the key point being that (?<=a) asserts that the "immediately preceding
> character cannot be 'a'
>
> Am I misreading this? I'd expect that to be the meaning of (?<!a),
> negative lookbehind.
>

No, typo on my behalf. The problem exists on any form of negative
lookbehind variable length alternation, positive or negative.


> I actually find it questionable if it should be allowed at all. I think it
> wasn't once upon a time.
>
> Experimental support for variable-length lookbehind was introduced in
> 5.30, it was a fatal error before that. I stumbled on this issue when
> Ricardo proposed making it non-experimental.
>
It is *definitelty* not suitable for marking as non-experimental. It was
good of you to find this.


> Anyway, for now the only solution is to teach regcomp.c to not trie inside
> of lookbehind, or to perhaps not trie if the keys are different lengths,
> probably the former is easier to implement than the latter so that is what
> I will do.
>
> I don't think my own use of lookbehinds would lose much from the lack of
> trie-optimization - I rarely have anything particularly complex in there -
> but I don't know how other people's usage would stack up. But it would seem
> a shame if a side-effect of this experimental feature is that we have to
> pessimize what we supported before - @rjbs <https://github.com/rjbs>,
> @khwilliamson <https://github.com/khwilliamson> do you have thoughts on
> this?
>

Unfortunately it is not the trie optimization at fault. Even if I disable
it inside of lookbehind the bug persists. It is inherent to the problem
that (?<= x | yy ) and (?! x | yy) are not really equivalent to "a or b",
but are actually equivalent "a" and "b", and thus to unrolling the branches
and putting them in their own assertions.

The work around is to change (?<= x | yy) into (?<= x) (?<= yy).

Maybe I can find a fix, I am still digging.

cheers,
yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by hvds** [comment]
> The work around is to change (?<= x | yy) into (?<= x) (?<= yy).

I'm not seeing this - assuming `x`, `yy` are literals, the RHS can never match, but the LHS can match. Do you mean to change it to `(?: (?<=x) | (?<=yy) )`?

**Comment by demerphq** [comment]
On Sun, 20 Feb 2022 at 06:17, Hugo van der Sanden ***@***.***>
wrote:

> The work around is to change (?<= x | yy) into (?<= x) (?<= yy).
>
> I'm not seeing this - assuming x, yy are literals, the RHS can never
> match, but the LHS can match. Do you mean to change it to (?: (?<=x) |
> (?<=yy) )?
>
Sorry, I am muddling things up. The work around I showed only applies to
negative lookbehind, not positive.

The formulation you showed is the work around for positive lookbehind.

My apologies.

Yves

**Comment by demerphq** [comment]
On Sun, 20 Feb 2022 at 06:36, demerphq ***@***.***> wrote:

> On Sun, 20 Feb 2022 at 06:17, Hugo van der Sanden <
> ***@***.***> wrote:
>
>> The work around is to change (?<= x | yy) into (?<= x) (?<= yy).
>>
>> I'm not seeing this - assuming x, yy are literals, the RHS can never
>> match, but the LHS can match. Do you mean to change it to (?: (?<=x) |
>> (?<=yy) )?
>>
> Sorry, I am muddling things up. The work around I showed only applies to
> negative lookbehind, not positive.
>
> The formulation you showed is the work around for positive lookbehind.
>
> My apologies.
>
>
Well, I have been working on this all day without much progress. Part of
the problem I am facing is that the SUCCEED opcode is used by multiple ops
besides the IFMATCH and UNLESSM, but we want special logic to happen when
we finish a lookbehind operation. I may have to create a new opcode for
this purpose.

cheers,
Yves


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
On Sun, 20 Feb 2022 at 10:02, demerphq ***@***.***> wrote:

> On Sun, 20 Feb 2022 at 06:36, demerphq ***@***.***> wrote:
>
>> On Sun, 20 Feb 2022 at 06:17, Hugo van der Sanden <
>> ***@***.***> wrote:
>>
>>> The work around is to change (?<= x | yy) into (?<= x) (?<= yy).
>>>
>>> I'm not seeing this - assuming x, yy are literals, the RHS can never
>>> match, but the LHS can match. Do you mean to change it to (?: (?<=x) |
>>> (?<=yy) )?
>>>
>> Sorry, I am muddling things up. The work around I showed only applies to
>> negative lookbehind, not positive.
>>
>> The formulation you showed is the work around for positive lookbehind.
>>
>> My apologies.
>>
>>
> Well, I have been working on this all day without much progress. Part of
> the problem I am facing is that the SUCCEED opcode is used by multiple ops
> besides the IFMATCH and UNLESSM, but we want special logic to happen when
> we finish a lookbehind operation. I may have to create a new opcode for
> this purpose.
>
>
Fixed in

https://github.com/Perl/perl5/pull/19442

More tests would be nice.

cheers,
Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by khwilliamson** [comment]
If we can't close it, we should rename it to effectively "need more tests"

**Comment by demerphq** [comment]
I think we can close this, I added more related tests in the ACCEPT branch.
